GCC warnings in release mode

**Describe the bug**

Building in `Release` mode with GCC 11 spawns the following warnings:
```
[2/12] Building CXX object server/CMakeFiles/server.dir/edithand.cpp.o
/home/louis/freeciv/freeciv21/server/edithand.cpp: In function 'void handle_edit_game(connection*, const packet_edit_game*)':
/home/louis/freeciv/freeciv21/server/edithand.cpp:1421:17: warning: 'int strncmp(const char*, const char*, size_t)' specified bound 4096 exceeds source size 1365 [-Wstringop-overread]
 1421 |       != strncmp(packet->scenario_authors, game.scenario.authors,
      |          ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 1422 |                  MAX_LEN_PACKET)) {
      |                  ~~~~~~~~~~~~~~~
[3/12] Building CXX object tools/CMakeFiles/freeciv21-manual.dir/civmanual.cpp.o
/home/louis/freeciv/freeciv21/tools/civmanual.cpp: In function 'bool manual_command(tag_types*)':
/home/louis/freeciv/freeciv21/tools/civmanual.cpp:703:18: warning: 'doc' may be used uninitialized in this function [-Wmaybe-uninitialized]
  703 |           fprintf(doc, tag_info->item_begin, "tech", ptech->item_number);
      |           ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[4/12] Building CXX object server/CMakeFiles/server.dir/stdinhand.cpp.o
[...]
/home/louis/freeciv/freeciv21/server/stdinhand.cpp: In function 'bool playernation_command(connection*, char*, bool)':
/home/louis/freeciv/freeciv21/server/stdinhand.cpp:4299:22: warning: 'pstyle' may be used uninitialized in this function [-Wmaybe-uninitialized]
 4299 |       pplayer->style = pstyle;
      |       ~~~~~~~~~~~~~~~^~~~~~~~
[10/11] Building CXX object tools/ruledit/CMakeFiles/freeciv21-ruledit.dir/ruledit_qt.cpp.o
/home/louis/freeciv/freeciv21/tools/ruledit/ruledit_qt.cpp: In function 'void ruledit_qt_display_requirers(const char*, void*)':
/home/louis/freeciv/freeciv21/tools/ruledit/ruledit_qt.cpp:72:21: warning: 'this' pointer is null [-Wnonnull]
   72 |   gui->show_required(requirers, msg);
      |   ~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~
/home/louis/freeciv/freeciv21/tools/ruledit/ruledit_qt.cpp:271:6: note: in a call to non-static member function 'void ruledit_gui::show_required(requirers_dlg*, const char*)'
  271 | void ruledit_gui::show_required(requirers_dlg *requirers, const char *msg)
      |      ^~~~~~~~~~~
[11/11] Linking CXX executable freeciv21-ruledit
```
They look serious.

**To Reproduce**
Steps to reproduce the behavior:
1. Use GCC 11
2. Configure with `-DCMAKE_BUILD_TYPE=Release`
2. Build
3. See warnings

**Expected behavior**

No such warnings (only Qt deprecations).

**Platform and version (please complete the following information):**
 - OS: Arch Linux
 - Freeciv21 version: e9919bd18deb8355bdd2058fef3cff79ac5a9fa0

--- a/ai/default/daicity.cpp
+++ b/ai/default/daicity.cpp
@@ -343,9 +343,11 @@ static void dai_city_choose_build(struct ai_type *ait,
       break;
     case CT_NONE:
       build_new.kind = VUT_NONE;
+      build_new.value.age = 0; // Don't leave it uninitialized
       break;
     case CT_LAST:
       build_new.kind = universals_n_invalid();
+      build_new.value.age = 0; // Don't leave it uninitialized
       break;
     };
 

--- a/client/gui-qt/citydlg.cpp
+++ b/client/gui-qt/citydlg.cpp
@@ -1635,6 +1635,7 @@ void city_dialog::save_cma()
     QByteArray ask_bytes = ask->input_edit.text().toLocal8Bit();
     QString text = ask_bytes.data();
     if (!text.isEmpty()) {
+      param.max_growth = false;
       param.allow_disorder = false;
       param.allow_specialists = true;
       param.require_happy = ui.qsliderbox->cma_celeb_checkbox->isChecked();

--- a/client/gui-qt/mapctrl.cpp
+++ b/client/gui-qt/mapctrl.cpp
@@ -229,7 +229,7 @@ void map_view::shortcut_pressed(int key)
   if (bt == sc->mouse && md == sc->mod && king()->rallies.hover_city) {
     char text[1024];
 
-    if (tile_city(ptile)) {
+    if (ptile && tile_city(ptile)) {
       king()->rallies.hover_tile = true;
       king()->rallies.rally_city = tile_city(ptile);
 
@@ -267,7 +267,7 @@ void map_view::shortcut_pressed(int key)
     return;
   }
 
-  if (bt == Qt::LeftButton && king()->menu_bar->delayed_order) {
+  if (bt == Qt::LeftButton && king()->menu_bar->delayed_order && ptile) {
     king()->menu_bar->set_tile_for_order(ptile);
     clear_hover_state();
     exit_goto_state();
@@ -275,7 +275,7 @@ void map_view::shortcut_pressed(int key)
     return;
   }
 
-  if (bt == Qt::LeftButton && king()->menu_bar->quick_airlifting) {
+  if (bt == Qt::LeftButton && king()->menu_bar->quick_airlifting && ptile) {
     if (tile_city(ptile)) {
       multiairlift(tile_city(ptile), king()->menu_bar->airlift_type_id);
     } else {

--- a/client/options.cpp
+++ b/client/options.cpp
@@ -3814,6 +3814,7 @@ static void load_cma_preset(struct section_file *file, int i)
         secfile_lookup_int_default(file, 0, "cma.preset%d.factor%d", i, o);
   }
   output_type_iterate_end;
+  parameter.max_growth = false;
   parameter.require_happy =
       secfile_lookup_bool_default(file, false, "cma.preset%d.reqhappy", i);
   parameter.happy_factor =

--- a/common/city.cpp
+++ b/common/city.cpp
@@ -3430,7 +3430,7 @@ void destroy_city_virtual(struct city *pcity)
   if (pcity->tile_cache != NULL) {
     free(pcity->tile_cache); // realloc
   }
-  NFCPP_FREE(pcity->cm_parameter);
+  NFC_FREE(pcity->cm_parameter);
   if (!is_server()) {
     unit_list_destroy(pcity->client.info_units_supported);
     unit_list_destroy(pcity->client.info_units_present);

--- a/common/generate_packets.py
+++ b/common/generate_packets.py
@@ -290,29 +290,34 @@ class Field:
         if not self.is_array:
             return "  differ = (old->%(name)s != real_packet->%(name)s);" % self.__dict__
 
+        sizes = None, None
         if self.dataio_type == "string" or self.dataio_type == "estring":
             c = "strcmp(old->%(name)s[i], real_packet->%(name)s[i]) != 0" % self.__dict__
-            array_size_u = self.array_size1_u
-            array_size_o = self.array_size1_o
+            sizes = self.array_size1_o, self.array_size1_u
         elif self.is_struct:
             c = "!are_%(dataio_type)ss_equal(&old->%(name)s[i], &real_packet->%(name)s[i])" % self.__dict__
+            sizes = self.array_size_o, self.array_size_u
         else:
             c = "old->%(name)s[i] != real_packet->%(name)s[i]" % self.__dict__
+            sizes = self.array_size_o, self.array_size_u
 
-        return '''
-    {
-      differ = (%(array_size_o)s != %(array_size_u)s);
-      if (!differ) {
+        differ = f'({sizes[0]} != {sizes[1]})'
+        if sizes[0] == sizes[1]:
+          differ = 'false'
+        return f'''
+    {{
+      differ = {differ};
+      if (!differ) {{
         int i;
 
-        for (i = 0; i < %(array_size_u)s; i++) {
-          if (%(c)s) {
+        for (i = 0; i < {sizes[1]}; i++) {{
+          if ({c}) {{
             differ = true;
             break;
-          }
-        }
-      }
-    }''' % self.get_dict(vars())
+          }}
+        }}
+      }}
+    }}'''
 
     # Returns a code fragment which updates the bit of the this field
     # in the "fields" bitvector. The bit is either a "content-differs"

--- a/common/helpdata.cpp
+++ b/common/helpdata.cpp
@@ -1836,8 +1836,8 @@ char *helptext_unit(char *buf, size_t bufsz, struct player *pplayer,
   // Add requirement text for the unit type itself
   requirement_vector_iterate(&utype->build_reqs, preq)
   {
-    req_text_insert_nl(buf, bufsz, pplayer, preq, VERB_DEFAULT,
-                       Q_("?bullet:* "));
+    (void) req_text_insert_nl(buf, bufsz, pplayer, preq, VERB_DEFAULT,
+                              Q_("?bullet:* "));
   }
   requirement_vector_iterate_end;
 

--- a/common/path_finder.cpp
+++ b/common/path_finder.cpp
@@ -433,11 +433,6 @@ void path_finder::path_finder_private::attempt_paradrop(
     {
       if (distance >= action->min_distance * action->min_distance) {
         // Paradrop has many hard requirements, see unittools.cpp:do_paradrop
-        // The target tile must be known
-        if (target == nullptr) {
-          continue;
-        }
-
         // The target tile must be seen or be a native tile
         if (TILE_KNOWN_SEEN != tile_get_known(target, probe.owner)
             && !is_native_tile(probe.utype, target)) {

--- a/common/unit.cpp
+++ b/common/unit.cpp
@@ -1508,9 +1508,8 @@ struct unit *unit_virtual_create(struct player *pplayer, struct city *pcity,
   fc_assert_ret_val(NULL != punittype, NULL); // No untyped units!
   fc_assert_ret_val(NULL != pplayer, NULL);   // No unowned units!
 
-  /* Make sure that contents of unit structure are correctly initialized,
-   * if you ever allocate it by some other mean than fc_calloc() */
-  struct unit *punit = new unit[1]();
+  // Make sure that contents of unit structure are correctly initialized.
+  struct unit *punit = new unit();
   int max_vet_lvl;
 
   // It does not register the unit so the id is set to 0.
@@ -1551,6 +1550,8 @@ struct unit *unit_virtual_create(struct player *pplayer, struct city *pcity,
 
   punit->carrying = NULL;
 
+  punit->changed_from = ACTIVITY_IDLE;
+  punit->changed_from_count = 0;
   set_unit_activity(punit, ACTIVITY_IDLE);
   punit->battlegroup = BATTLEGROUP_NONE;
   punit->has_orders = false;

--- a/common/unit.cpp
+++ b/common/unit.cpp
@@ -1633,7 +1633,7 @@ void unit_virtual_destroy(struct unit *punit)
   }
 
   if (--punit->refcount <= 0) {
-    FCPP_FREE(punit);
+    FC_FREE(punit);
   }
 }
 

--- a/server/cityhand.cpp
+++ b/server/cityhand.cpp
@@ -586,7 +586,7 @@ void handle_city_manager(struct player *pplayer, int city_id, bool enabled,
   }
 
   if (!pcity->cm_parameter) {
-    pcity->cm_parameter = new cm_parameter[1]();
+    pcity->cm_parameter = new cm_parameter();
   }
 
   cm_copy_parameter(pcity->cm_parameter, &parameter);

--- a/server/citytools.cpp
+++ b/server/citytools.cpp
@@ -1255,7 +1255,7 @@ bool transfer_city(struct player *ptaker, struct city *pcity,
 
     // Clear CMA.
     if (pcity->cm_parameter) {
-      free(pcity->cm_parameter);
+      delete pcity->cm_parameter;
       pcity->cm_parameter = NULL;
     }
 

--- a/server/cityturn.cpp
+++ b/server/cityturn.cpp
@@ -409,7 +409,7 @@ void auto_arrange_workers(struct city *pcity)
   if (!cmr->found_a_valid) {
     if (pcity->cm_parameter) {
       // If player-defined parameters fail, cancel and notify player.
-      free(pcity->cm_parameter);
+      delete pcity->cm_parameter;
       pcity->cm_parameter = NULL;
 
       notify_player(city_owner(pcity), city_tile(pcity), E_CITY_CMA_RELEASE,

--- a/server/edithand.cpp
+++ b/server/edithand.cpp
@@ -1419,7 +1419,7 @@ void handle_edit_game(struct connection *pc,
 
   if (0
       != strncmp(packet->scenario_authors, game.scenario.authors,
-                 MAX_LEN_PACKET)) {
+                 sizeof(packet->scenario_authors))) {
     sz_strlcpy(game.scenario.authors, packet->scenario_authors);
     changed = true;
   }

--- a/server/stdinhand.cpp
+++ b/server/stdinhand.cpp
@@ -4294,6 +4294,8 @@ static bool playernation_command(struct connection *caller, char *str,
       }
     } else if (pnation != NO_NATION_SELECTED) {
       pstyle = style_of_nation(pnation);
+    } else {
+      pstyle = nullptr;
     }
 
     if (!check) {

--- a/tools/civmanual.cpp
+++ b/tools/civmanual.cpp
@@ -184,7 +184,7 @@ static QString ruleset;
  */
 static bool manual_command(struct tag_types *tag_info)
 {
-  FILE *doc;
+  FILE *doc = nullptr;
   char filename[40];
   struct connection my_conn;
 

--- a/tools/ruledit/ruledit_qt.cpp
+++ b/tools/ruledit/ruledit_qt.cpp
@@ -177,6 +177,9 @@ ruledit_gui::ruledit_gui(ruledit_main *main) : QObject(main)
   req_edits = req_edit_list_new();
   this->req_vec_fixers = req_vec_fix_list_new();
   effect_edits = effect_edit_list_new();
+
+  // FIXME Should get rid of the static variable.
+  gui = this;
 }
 
 /**

